File: fortran90/day5_part2_2017.f90
program maze_of_twisty_trampolines
    implicit none
    integer, allocatable :: offsets(:)
    integer :: i, steps, n, jump
    character(len=100) :: line
    integer :: unit

    ! Open the input file
    unit = 10
    open(unit, file='input.txt', status='old', action='read')

    ! Read the offsets from the file
    n = 0
    do while (.true.)
        read(unit, '(A)', iostat=i) line
        if (i /= 0) exit
        n = n + 1
    end do
    rewind(unit)

    allocate(offsets(n))
    do i = 1, n
        read(unit, *) offsets(i)
    end do
    close(unit)

    ! Process the jumps
    i = 1
    steps = 0
    do while (i > 0 .and. i <= n)
        jump = offsets(i)
        offsets(i) = offsets(i) + 1
        if (jump >= 3) then
            offsets(i) = offsets(i) - 2
        end if
        i = i + jump
        steps = steps + 1
    end do

    print *, steps
end program maze_of_twisty_trampolines

File: pascal/day5_part2_2017.pas
program MazeOfTwistyTrampolines;

{$APPTYPE CONSOLE}
uses
  SysUtils;

var
  inFile: Text;
  baseOffsets: array of Integer;
  offsets1, offsets2: array of Integer;
  n, i: Integer;
  pos: Integer;
  steps1, steps2: Int64;
  oldOffset: Integer;
  line: String;
begin
  { Read all offsets from input.txt into baseOffsets }
  Assign(inFile, 'input.txt');
  Reset(inFile);
  while not Eof(inFile) do
  begin
    ReadLn(inFile, line);
    if line = '' then Continue;
    SetLength(baseOffsets, Length(baseOffsets) + 1);
    baseOffsets[High(baseOffsets)] := StrToInt(Trim(line));
  end;
  Close(inFile);

  { Prepare two copies for part 1 and part 2 }
  n := Length(baseOffsets);
  SetLength(offsets1, n);
  SetLength(offsets2, n);
  for i := 0 to n - 1 do
  begin
    offsets1[i] := baseOffsets[i];
    offsets2[i] := baseOffsets[i];
  end;

  { Part 1: increment every jumped offset by 1 }
  pos := 0;
  steps1 := 0;
  while (pos >= 0) and (pos < n) do
  begin
    oldOffset := offsets1[pos];
    Inc(offsets1[pos]);
    Inc(pos, oldOffset);
    Inc(steps1);
  end;

  { Part 2: offsets >= 3 are decremented, otherwise incremented }
  pos := 0;
  steps2 := 0;
  while (pos >= 0) and (pos < n) do
  begin
    oldOffset := offsets2[pos];
    if oldOffset >= 3 then
      Dec(offsets2[pos])
    else
      Inc(offsets2[pos]);
    Inc(pos, oldOffset);
    Inc(steps2);
  end;

  { Output results }
  WriteLn('Part 1 steps: ', steps1);
  WriteLn('Part 2 steps: ', steps2);
end.